// react-filters/dist/AutoComplete/Tag.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _autoBind = require('../utils/autoBind');

var _autoBind2 = _interopRequireDefault(_autoBind);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Tag = function (_Component) {
  _inherits(Tag, _Component);

  function Tag(props) {
    _classCallCheck(this, Tag);

    var _this = _possibleConstructorReturn(this, (Tag.__proto__ || Object.getPrototypeOf(Tag)).call(this, props));

    (0, _autoBind2.default)(['handleRemove'], _this);
    return _this;
  }

  _createClass(Tag, [{
    key: 'handleRemove',
    value: function handleRemove() {
      this.props.onRemove({
        id: this.props.id,
        text: this.props.text
      });
    }
  }, {
    key: 'render',
    value: function render() {
      var _props = this.props,
          showRemove = _props.showRemove,
          id = _props.id,
          text = _props.text;

      return _react2.default.createElement(
        'div',
        { className: 'ac-tag', 'data-id': id },
        _react2.default.createElement(
          'span',
          { className: 'ac-tag-text' },
          text
        ),
        showRemove && _react2.default.createElement(
          'span',
          {
            className: 'ac-tag-remove',
            onClick: this.handleRemove
          },
          '\u2715'
        )
      );
    }
  }]);

  return Tag;
}(_react.Component);

exports.default = Tag;


Tag.propTypes = {
  id: _react.PropTypes.oneOfType([_react.PropTypes.number, _react.PropTypes.string]),
  onRemove: _react.PropTypes.func,
  showRemove: _react.PropTypes.bool,
  text: _react.PropTypes.string.isRequired
};
